### PR TITLE
Add --overwrite and --organize-files flags

### PR DESCRIPTION
**Pull request by zurk**

### CODE REVIEW

File: ast2vec/__main__.py [hunk 1]
```diff
@@ -72,6 +72,12 @@ def main():
         help="Output path where the .asdf will be stored.")
     repo2nbow_parser.add_argument("--gcs", default=None, dest="gcs_bucket",
                                   help="GCS bucket to use.")
+    group = repo2nbow_parser.add_argument_group("override")
+    group_ex = group.add_mutually_exclusive_group(required=False)
+    group_ex.add_argument("--no_override", action="store_false", dest="override_existing",
```
Comment (vmarkovtsev) [inline ast2vec/__main__.py]: Nope. One flag is enough. `--disable-overwrite`, by default is false.

File: ast2vec/__main__.py [hunk 2]
```diff
@@ -72,6 +72,12 @@ def main():
...
+                          help="Specify if you want to rewrite existing models (default)")
+    group_ex.add_argument("--override", action="store_true", dest="override_existing",
```
Comment (vmarkovtsev) [inline ast2vec/__main__.py]: Did you mean overwrite? override is a completely different thing

File: ast2vec/repo2/base.py [hunk 3]
```diff
@@ -34,14 +35,28 @@ class Repo2Base(PickleableLogger):
     MAX_FILE_SIZE = 200000
 
     def __init__(self, tempdir=None, linguist=None, log_level=logging.INFO,
-                 bblfsh_endpoint=None, timeout=DEFAULT_BBLFSH_TIMEOUT):
+                 bblfsh_endpoint=None, timeout=DEFAULT_BBLFSH_TIMEOUT, override_existing=True):
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: same here

File: ast2vec/repo2/base.py [hunk 4]
```diff
@@ -34,14 +35,28 @@ class Repo2Base(PickleableLogger):
...
+        """
+        Initializer of Repo2Base class
+        :param tempdir: If you will clone repositories they will be stored in tempdir
+        :param linguist: Path to linguist.
+        :param log_level: Log level of Repo2Base
+        :param bblfsh_endpoint: bblfsh server endpoint
+        :param timeout: timeout for bblfsh
+        :param override_existing: Rewrite existing models or skip them
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: same here

File: ast2vec/repo2/base.py [hunk 5]
```diff
@@ -34,14 +35,28 @@ class Repo2Base(PickleableLogger):
...
+        """
         super(Repo2Base, self).__init__(log_level=log_level)
         self._tempdir = tempdir
         self._cloner = RepoCloner(redownload=True, log_level=log_level)
         self._cloner.find_linguist(linguist)
         self._bblfsh = [BblfshClient(bblfsh_endpoint or "0.0.0.0:9432")
                         for _ in range(multiprocessing.cpu_count())]
         self._timeout = timeout
+        self._override_existing = override_existing
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: same here

File: ast2vec/repo2/base.py [hunk 6]
```diff
@@ -34,14 +35,28 @@ class Repo2Base(PickleableLogger):
...
+
+    @property
+    def override_existing(self):
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: ...you got it

File: ast2vec/repo2/base.py [hunk 7]
```diff
@@ -148,8 +163,8 @@ def _bblfsh_parse(self, thread_index, filepath, language):
         try:
             return self._bblfsh[thread_index].parse(
                 filepath, language=language, timeout=self._timeout)
-        except DecodeError as e:
-            msg = "bblfsh raised an DecodeError exception. Probably your protobuf is <= v3.3.2 " \
+        except RpcError as e:
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: Why? It is exactly DecodeError on macs with a wrong protobuf. If you want smth else, add it to the exception tuple.

File: ast2vec/repo2/base.py [hunk 8]
```diff
@@ -171,10 +186,18 @@ class RepoTransformer(Transformer):
     WORKER_CLASS = None
     DEFAULT_NUM_PROCESSES = 2
 
-    def __init__(self, num_processes=DEFAULT_NUM_PROCESSES, **args):
+    def __init__(self, num_processes=DEFAULT_NUM_PROCESSES, decompose_output=False, **args):
+        """
+        Base class for transformers from repository to WORKER_CLASS model
+        :param num_processes: Number of parallel processes to transform
+        :param decompose_output: Expand output path by subfolders using the first char of \
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: I don't like the name... `organize_files`

\+ make it an integer which sets the number of levels - not only the single one. By default is 0

File: ast2vec/repo2/base.py [hunk 9]
```diff
@@ -251,18 +281,34 @@ def process_repo(self, url_or_path, output) -> bool:
         :return: True if the operation was successful; otherwise, False.
         """
         repo2 = self.WORKER_CLASS(**self._args)
+        if os.path.exists(output):
+            if repo2.override_existing:
+                self._log.info("Model %s already exists, but will be override. If you want to "
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: overwrite

File: ast2vec/repo2/base.py [hunk 10]
```diff
@@ -251,18 +281,34 @@ def process_repo(self, url_or_path, output) -> bool:
...
         try:
             result = repo2.convert_repository(url_or_path)
             for proto in ("https://", "http://", "git://", "ssh://"):
                 if url_or_path.startswith(proto):
                     url_or_path = url_or_path.replace(proto, "")
             model = self.WORKER_CLASS.MODEL_CLASS()
+            self._log.info("Construct %s model...", url_or_path)
+            start = time()
             model.construct(**self.result_to_model_kwargs(result, url_or_path))
+            self._log.info("Construct %s model is done. time: %f", url_or_path, time() - start)
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: All timings must be activated under DEBUG and the code executed if the logger.isEnabledFor

File: ast2vec/repo2/base.py [hunk 11]
```diff
@@ -306,8 +352,8 @@ def transform(self, repos, output, num_processes=None):
         def process_repos():
             with multiprocessing.Pool(processes=num_processes) as pool:
                 pool.starmap(type(self).process_entry,
-                             zip(inputs, repeat(self._args), repeat(output), repeat(queue)))
-
+                             zip(inputs, repeat(self._args), repeat(output), repeat(queue),
+                                 repeat(self._decompose_output)))
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: You do not need it, remove. args contain this flag.

File: ast2vec/tests/test_resolve_symlink.py [hunk 12]
```diff
@@ -23,6 +27,9 @@ def readlink(path):
         with self.assertRaises(resolve_symlink.DanglingSymlinkError) as _:
             resolve_symlink.resolve_symlink("xxx")
 
+        os.path.exists = exists_
```
Comment (vmarkovtsev) [inline ast2vec/tests/test_resolve_symlink.py]: Always do such things under `try...finally`

File: ast2vec/uast.py [hunk 13]
```diff
@@ -23,7 +22,14 @@ def construct(self, filenames, uasts):
 
     def _load_tree_kwargs(self, tree):
         return dict(filenames=split_strings(tree["filenames"]),
-                    uasts=[ParseResponse.FromString(x) for x in tree["uasts"]])
+                    uasts=[type(self).parse_bblfsh_response(x) for x in tree["uasts"]])
```
Comment (vmarkovtsev) [inline ast2vec/uast.py]: Extract this into the separate PR so that I can merge it fast.

**Comment by zurk** [comment]
I will fix tests (try: ... finally: ...) and add more in my next pull request.

File: ast2vec/repo2/base.py [hunk 14]
```diff
@@ -1,4 +1,5 @@
 from collections import namedtuple
+from itertools import accumulate
 from itertools import repeat
```
Comment (vmarkovtsev) [inline ast2vec/repo2/base.py]: Duplicate itertools import

**Pull request merged**